### ISSUE TITLE
A question about using curl to access resource server?

### ISSUE THREAD

**Issue by wldandan** [kakawait/uaa-behind-zuul-sample#13]
Hi kakawait
Thanks for the awesome sample. It helps me a lot about understanding zuul with security.

Now I have one problem about the sample, could you help me? thank you very much.

### a). I can get token from zuul server by using curl, that is result expected.
`curl --insecure -H "Authorization: Basic $(echo -n 'acme:acmesecret' | base64)" http://localhost:8765/uaa/oauth/token -d grant_type=password -d username=user -d password=password -v
`
the token is 
`eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE0Nzc4NzEwMDAsInVzZXJfbmFtZSI6InVzZXIiLCJhdXRob3JpdGllcyI6WyJST0xFX0FETUlOIiwiUk9MRV9VU0VSIl0sImp0aSI6IjhiODdhNTNhLWU0MmEtNDkyYS05ZWMxLTY2YWE4NmU4ZDkxOCIsImNsaWVudF9pZCI6ImFjbWUiLCJzY29wZSI6WyJvcGVuaWQiXX0.TWef-agKnnKaWDiW44ROxIseL0tVG7TXZWYTiMIV4RoKVLXXCFsFb5l24gRjxRCZAhmyB-5N6l-axLflgvKEbLrPHYJj054nEE9bi_I5mi9b2y4Z2mII6sfMkjeTAlrlyGKN-s-oU-nO2tws-vKCAYtycKmTL_xsNMH3Qmm1-4haG9SZIFq4YJkC93RvBDCGxit7jV62XrEyDj1atPWQ8QJeJmp0I8z04sVKlxulzRYpkkFMKjvWAx5lgnx5B6m-0jhXgi8g4kejUpZK9yojDTNRKMkYJuhPyTVCLsakpTl3vwOBZ5xpO-o2mgd9yn-zvjo7P1TK1rFys9xvREu2mw`

### b). then, when I request zuul server by the command
`curl -v -H 'Content-type: application/json' -H 'Authorization: Bearer ACCESS_TOKEN' http://localhost:8765/dummy`
 the result always show 
" X-Frame-Options: DENY
< Location: http://localhost:8765/login"

for the case, how can I access the http://localhost:8765/dummy by curl with token rather than using browser? suppose that I have a client which can send api request to zuul server.

**Comment by wldandan** [comment]
<img width="984" alt="login-prompt" src="https://cloud.githubusercontent.com/assets/311428/19838111/1a496b60-9f03-11e6-8e46-cce7668c6bb6.png">

**Comment by kakawait** [comment]
@wldandan good point. By using `@EnableOauth2Sso` on `Zuul` (here *api-gateway*) you transform *api-gateway* as token relay but not as resource server.

So when fetching microservices using `Authorization` header *api-gateway* will not try to read this header, so it will consider you as not auth.

You should transform your *api-gateway* as `ResourceServer` (`@EnableResourceServer`) in order to add `OAuth2AuthenticationProcessingFilter` (in addition to `OAuth2ClientAuthenticationProcessingFilter` used by SSO) that will handle `Authorization` header.

I just created a branch to POC your needs. I have to polish it before merging on `master` so be careful if you choose to copy past.

---

In addition you could up-vote http://stackoverflow.com/questions/37059128/spring-1-3-3-enableresourceserver-and-enableoauth2sso-at-the-same-time (not my question) and up-vote https://github.com/spring-projects/spring-security-oauth/issues/683 because there is a trick to do for combining `@EnableResourceServer` and `@EnableOauth2Sso`

**Comment by kakawait** [comment]
@wldandan does the solution works for you ?

**Comment by wldandan** [comment]
@kakawait Hey kakawait, thanks for your solution. I did not get chance to verify it.
I will give it a try and respond you as soon as possible. :)
Thanks a lot.

**Comment by wldandan** [comment]
@kakawait Hey kakawait, thanks for your solution. I have verified the solution in 'resource-server' branch. It does work, thanks for your work, it is awesome :)

**Comment by kakawait** [comment]
Cool, I will keep issue open until I polish the code

**Comment by mavogel** [comment]
@kakawait first of all: awesome work! It helps me a lot. 

**Problem**
It have the same issue like wldadan with adding a client who uses the zuul proxy, but I'd like to have multiple users accessing the resourced by the `grand_type=password`

1. I removed the user and password property from `application.yml` of the `uaa-service` 
2. I added more users for authentication
```java
@Override
        protected void configure(AuthenticationManagerBuilder auth) throws Exception {
            auth
                    .inMemoryAuthentication()
                    .withUser("user").password("password")
                    .authorities("USER").roles("USER")
                    .and()
                    .withUser("admin").password("admin")
                    .authorities("ADMIN").roles("ADMIN");
//            auth.parentAuthenticationManager(authenticationManager);
        }
```

It works fine in the browser with `admin` credentials, but `curl --insecure -H "Authorization: Basic $(echo -n 'acme:acmesecret' | base64)" http://localhost:8765/uaa/oauth/token -d grant_type=password -d username=admin -d password=admin -v`
results in
```bash
*   Trying ::1...
* Connected to localhost (::1) port 8765 (#0)
> POST /uaa/oauth/token HTTP/1.1
> Host: localhost:8765
> User-Agent: curl/7.43.0
> Accept: */*
> Authorization: Basic YWNtZTphY21lc2VjcmV0
> Content-Length: 49
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 49 out of 49 bytes
< HTTP/1.1 400
< Set-Cookie: JSESSIONID=85A5C7AD4F55891EF9F0EA56BCF4087A;path=/;HttpOnly
< Set-Cookie: XSRF-TOKEN=2b264f92-1cb6-4a1c-9f99-6fc25f66546c;path=/;Secure
< X-Application-Context: api-gateway:docker:8765
< Date: Mon, 09 Jan 2017 12:12:13 GMT
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
< Content-Type: application/json;charset=UTF-8
< Transfer-Encoding: chunked
< Connection: close
<
* Closing connection 0
{"error":"invalid_grant","error_description":"Bad credentials"}
```

With the log message: `o.s.s.a.dao.DaoAuthenticationProvider    : User 'admin' not found`. The same happens with `user` and `password`.

**Possible idea to solve:**
My idea now (which I didn't manage to implement) to share the authentication manager of the `WebSecurityAdapter` with the `AuthorizationServerConfigurerAdapter`. 

```java
 @Override
        public void configure(AuthorizationServerEndpointsConfigurer endpoints) throws Exception {
            // TODO here
            endpoints.authenticationManager(authenticationManager).accessTokenConverter(jwtAccessTokenConverter());
        }
```

I made a [branch](https://github.com/mavogel/uaa-behind-zuul-sample/tree/integration-resource-server) for it in my fork.

=> Do you know how to do that? Or is there another (better) way? Thanks in advance!

**Comment by mavogel** [comment]
The solution was 
1. Exposing the AuthenticationManager from `LoginConfiguration`
```java
@Override
@Bean
public AuthenticationManager authenticationManagerBean() throws Exception {
    return super.authenticationManagerBean();
}
```

2. Injecting it in the `AuthorizationServerConfiguration`
```java
@Autowired
@Qualifier("authenticationManagerBean")
private AuthenticationManager authenticationManager;
```

3. I also added a protected resource only accessible for the `ROLE_ADMIN` in the `DummyServiceApplication` which gets an `ACCESS_DENIED` for the `user`:
```java
@PreAuthorize("#oauth2.hasScope('openid') and hasRole('ROLE_ADMIN')")
@RequestMapping(value = "secret", method = RequestMethod.GET)
@ResponseBody
public String helloSecret(Principal principal) {
    return principal == null ? "Hello anonymous" : "S3CR3T  - Hello " + principal.getName();
}
```

Props to [PiggyMetrics](https://github.com/sqshq/PiggyMetrics/) who brought me up to the solution and you guys for this awesome work!

**Comment by kakawait** [comment]
Sorry I didn't have time to read your question. I will check your question and solution

But thank you for the solution

**Comment by kentoj** [comment]
@mavogel I followed your fix and I got the token, but how do I use it?
I tried `curl http://localhost:8765/dummy/secret -H "Authorization: Bearer ${TOKEN}"` and I did not get anything back.

when I do a -v output on curl I get the following:
```
➜  blog-microservices git:(B3) curl http://localhost:8765/dummy/secret -H "Authorization: Bearer ${TOKEN}" -v
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8765 (#0)
> GET /dummy/secret HTTP/1.1
> Host: localhost:8765
> User-Agent: curl/7.51.0
> Accept: */*
> Authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE0OTIwMDY3MzksInVzZXJfbmFtZSI6ImFkbWluIiwiYXV0aG9yaXRpZXMiOlsiUk9MRV9BRE1JTiJdLCJqdGkiOiJkZjg5MGZhZi05NzJkLTQ5OWUtOWNhNy0zMjM5NjA2NzE1OWYiLCJjbGllbnRfaWQiOiJhY21lIiwic2NvcGUiOlsib3BlbmlkIl19.QlwydyvMesEEuCx7cJJ0sGVvT3y3kMQ25_H_LDuAwDDz1m87_KmMDWeEWF6CblK6TI9-__bz8V_38oh4_qTR7icf4MMvtmcFOBCnK_0LkAl4fnO-7tzOjhTeEZdbJ0zQFp81ZpI0LiM4_m7sbWuvFL7Sx6hHwVFel-1SKjfltTTUCXYFoKxgaMOSwvRJDoT209nNntvQAPnqCKqEuKZmawFbNHzhrtKmqEXgBqGmaohn8otONfpWwwG79Jn_duW79H3wpH4VedxIGeGVs4OsdBR5fC2MDQWP2oq8NDF7XCeFYspYU9apECioTj42fE48w7RzRia26ceMVC8xMRVRnQ
>
< HTTP/1.1 302
< Set-Cookie: JSESSIONID=6E688E080CD119409E022E4828FCBCEB;path=/;HttpOnly
< Set-Cookie: XSRF-TOKEN=6b953f9a-88fe-4e05-9665-ad21fa565898;path=/;Secure
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
< Location: http://localhost:8765/login
< Content-Length: 0
< Date: Wed, 12 Apr 2017 02:28:23 GMT
<
* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact

```

It looks like I am being forwarded to the /login endpoint so I must be missing a header or something. How do you use the OAuth2 token gotten through CURL?

**Comment by kentoj** [comment]
This solution in the resource-server branch worked for me also. Is there a reason it wouldn't be merged to master? I will submit a PR for this if it is wanted in master.

**Comment by mavogel** [comment]
Yep @kakawait, I think you could merge it into the master branch because the `resource-server` branch is an enhancement of the master atm.

**Comment by kentoj** [comment]
@mavogel He accepted my PR: https://github.com/kakawait/uaa-behind-zuul-sample/pull/23
